### ISSUE TITLE
Regex failing to parse valid JS regular expression, even when RegexOptions.ECMAScript is specified

### ISSUE THREAD

**Issue by madelson** [dotnet/corefx#41546]
The following regular expression works in JavaScript (at least in Chrome and IE): `/[^]/`.

However, this same pattern causes an `ArgumentException` when passed to .NET's regex class in ECMAScript compliance mode:

```
var regex = new Regex("[^]", RegexOptions.ECMAScript);
```

> parsing "[^]" - Unterminated [] set.

**Comment by danmosemsft** [comment]
Related https://github.com/dotnet/corefx/issues/31786

**Comment by pgovind** [comment]
I ran into this while looking at #31786. They differ slightly but the bug exists in `RegexParser.ScanCharClass` in both cases.

Description:
As part of initializing our RegEx, we call `CountCaptures` -> Consume the 0th character (`[` in this case) -> `ScanCharClass(caseInsensitive: false, scanOnly: true)`. In `ScanCharClass`, we consume `^` -> move ahead by 1 -> Consume `]`. The bug exists after this. The local bool `firstChar` is always initialized to `false` in `ScanCharClass`. Once we consume `]`, there's code to mark the set as `closed` ONLY when `firstChar` is false. Once we run out of input, we check that `closed` is true and throw if not. That's the bug. It seems that we should maybe rely on `_currentPosition` instead of the local `firstChar` to detect how many characters we have parsed, but I want to be careful of making changes here. I'll do some more code review before coming up with a solution.
